src/main/java/com/pusher/client/channel/impl/ChannelManager.java

```diff
@@ -73,9 +73,11 @@ public void setConnection(final InternalConnection connection) {
     public void subscribeTo(final InternalChannel channel, final ChannelEventListener listener,
             final String... eventNames) {
 
-        validateArgumentsAndBindEvents(channel, listener, eventNames);
-        channelNameToChannelMap.put(channel.getName(), channel);
-        sendOrQueueSubscribeMessage(channel);
+        synchronized (channelNameToChannelMap){
+            validateArgumentsAndBindEvents(channel, listener, eventNames);
```

Comment:
Do we need these non-map accesses to be inside the critical section? Some of them may also take locks...